client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/selectors.js

```diff
@@ -0,0 +1,69 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import { get, isEmpty } from 'lodash';
+import { translate } from 'i18n-calypso';
+/**
+ * Internal dependencies
+ */
+import { getSelectedSiteId } from 'state/ui/selectors';
+
+export const getCarrierAccountsState = ( state, siteId = getSelectedSiteId( state ), carrier ) => {
+	return get(
+		state,
+		[ 'extensions', 'woocommerce', 'woocommerceServices', siteId, 'carrierAccounts', carrier ],
+		null
+	);
+};
+
+const emailRegex = /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/;
+
+const urlRegex = /^(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z0-9\u00a1-\uffff][a-z0-9\u00a1-\uffff_-]{0,62})?[a-z0-9\u00a1-\uffff]\.)+(?:[a-z\u00a1-\uffff]{2,}\.?))(?::\d{2,5})?(?:[/?#]\S*)?$/i;
+
+export const getFormErrors = ( state, siteId, carrier ) => {
+	const carrierAccountState = getCarrierAccountsState( state, siteId, carrier );
+	const { fieldErrors, ignoreValidation, requiredFields, values } = carrierAccountState.settings;
+
+	if ( isEmpty( values ) ) {
+		return {};
+	}
+
+	for (const field of requiredFields) {
+		const value = values[ field ];
+		if ( ! value || ( 'string' === typeof value && '' === value.trim() ) ) {
+			fieldErrors[ field ] =  translate( 'This field is required' );
+		} else {
+			delete fieldErrors[ field ];
+		}
+	}
+
+	if ( fieldErrors.license_agreement ) {
+		fieldErrors.license_agreement = translate( 'You have to accept the license agreement' );
+	}
+
+	if ( values.email  && ! values.email.match( emailRegex ) ) {
+		fieldErrors.email = translate( "The email format is not valid" );
+	}
+	if ( values.company_website  && ! values.company_website.match( urlRegex ) ) {
+		fieldErrors.company_website = translate( "The company website format is not valid" );
+	}
+
+	if ( ignoreValidation ) {
```

Comment:
If we leave fields blank, `fieldErrors` will be an object with key = field_name, and value = `'This field is required'`.  But since all validation are ignored except country, the `'This field is required'` error will be removed. 

I think the `'This field is required'` should stay without being removed.

Reply:
hey @harriswong why do you say validation is ignored except for the country? It's actually the other way around, at first validation is only ignored for country (because it's the only field starting with value)

Reply:
Because `ignoreValidation` is only `false` for `country`
![image](https://user-images.githubusercontent.com/572862/87358287-b84f8300-c522-11ea-9da2-dc381527b0d1.png)

Reply:
Oh I see what you mean by the "first validation". On page load, the `fieldErrors` are all `This field is required`:
![image](https://user-images.githubusercontent.com/572862/87360952-66a9f700-c528-11ea-818e-41e19fc99b20.png)

But `ignoreValidations` values are all set to `true` except `country` so that all the `fieldsErrors` are removed on page load, such that there is no "This field is required" showing initially?

Shouldn't we show `This field is required` initially because they are required?

Reply:
> Shouldn't we show This field is required initially because they are required?
@elizaan36 , correct me if I'm wrong, but I think the idea was to only display the required field error message after the user updates the field and leaves it blank. The optional fields have an "optional" keyword in their label, so the user is supposed to fill all the other ones. That's what I understood from the prototype.

Reply:
> only display the required field error message after the user updates the field and leaves it blank.

That's correct, @Ferdev !

Reply:
Noticed some things as I was testing the error state messaging that would help people complete the form easier, somewhat along the lines of what @harriswong suggested except only after the user has interacted with the field. We want to avoid the possibility of someone opening the form full of errors before they've started.

For the UPS account number error message (the first field), if the user enters less than 6 digits in the field the error message could say "Your UPS account number needs to be 6 digits in length" or something like that. Currently "This field is required" doesn't provide enough information in case they've entered something incorrectly.

For the Zip code and phone number fields, we could also give them a "The zip code format is not valid" error message for anything that's not 5 digits, and "The phone number format is not valid" for anything less than 10 digits.